File: src/engine/adapters/redis/mutate.ts
import type { MutationPlan, MutationResult, MutationRowOp } from '@shared/domain/mutations';
import { encodePath } from '@shared/domain/tree';
import type { Redis } from 'ioredis';
import type { OpCtx } from '../adapter';
import { AdapterError, assertWritable, throwIfCancelled } from '../errors';
import { mapError } from './errors';

// The reserved sentinels for redis mutations, expressed through the existing relational-shaped
// MutationRowOp rather than widening the shared mutation schema — mirrors mongo/mutate.ts's
// `$document` precedent. `_key` names the target redis key: `plan.path` only ever resolves to a
// database (never a specific key, unlike a SQL table row or a Mongo `_id`), so every op has to
// carry its own key name — including `insert`, which by definition has no existing key a path
// could point at yet. `$value` carries a string SET's new value; `$` can never start a real
// redis key's *field* name inside a hash and `_key` is never a field name at all here (every op
// this module accepts has exactly one field, one of these two), so neither can collide with
// genuine data.
const KEY_SENTINEL = '_key';
const VALUE_SENTINEL = '$value';

function resolveDatabaseSegment(path: MutationPlan['path']): string {
  const [databaseSegment] = path.segments;
  if (databaseSegment?.kind !== 'database') {
    throw new AdapterError(
      'E_NOT_FOUND',
      `mutate requires a database-rooted path, got: ${encodePath(path.segments)}`,
    );
  }
  return databaseSegment.name;
}

function keyNameFrom(values: Record<string, string | null>, label: string): string {
  const raw = values[KEY_SENTINEL];
  if (typeof raw !== 'string' || raw === '') {
    throw new AdapterError(
      'E_QUERY',
      `a redis ${label} mutation requires a non-empty ${KEY_SENTINEL}`,
    );
  }
  return raw;
}

function valueFrom(values: Record<string, string | null>, label: string): string {
  const raw = values[VALUE_SENTINEL];
  if (typeof raw !== 'string') {
    throw new AdapterError(
      'E_UNSUPPORTED',
      `a redis ${label} mutation requires a ${VALUE_SENTINEL}`,
    );
  }
  return raw;
}

function renderOpText(op: MutationRowOp): string {
  if (op.kind === 'update')
    return `SET ${keyNameFrom(op.key, 'update')} ${valueFrom(op.changes, 'update')}`;
  if (op.kind === 'delete') return `DEL ${keyNameFrom(op.key, 'delete')}`;
  return `SET ${keyNameFrom(op.values, 'insert')} ${valueFrom(op.values, 'insert')} NX`;
}

// Synchronous (Adapter rule 3): no network, no TYPE lookup — trusts the plan's shape as given,
// same discipline as mongo/mutate.ts's preview().
export function preview(plan: MutationPlan): string[] {
  resolveDatabaseSegment(plan.path);
  return plan.ops.map(renderOpText);
}

// Edit is scoped to string-type keys only in this version (documented scope decision, mirrors
// how P8 scoped Mongo's insert out and P4 scoped DDL's structured metadata out): a hash/list/
// set/zset/stream each has its own per-element mutation semantics (HSET a field, LSET an index,
// SADD/SREM, ZADD, XADD) that are a materially bigger job than a single SET. The UI already
// disables the edit action for those types; this TYPE check enforces it server-side too (§8.12's
// standard — never only greyed out).
async function assertEditableType(conn: Redis, key: string): Promise<void> {
  let rawType: string;
  try {
    rawType = await conn.type(key);
  } catch (err) {
    throw mapError(err);
  }
  if (rawType !== 'none' && rawType !== 'string') {
    throw new AdapterError(
      'E_UNSUPPORTED',
      `only string-type keys are editable in this version, ${key} is ${rawType}`,
    );
  }
}

export async function mutate(
  conn: Redis,
  ctx: OpCtx,
  readOnly: boolean,
  plan: MutationPlan,
): Promise<MutationResult> {
  // §8.12's standard: enforced here, not only greyed out in the UI (mirrors mongo/mariadb/
  // mutate.ts).
  assertWritable(readOnly);
  resolveDatabaseSegment(plan.path);
  ctx.setCommand(preview(plan).join(';\n'));

  let affectedRows = 0;
  try {
    for (const op of plan.ops) {
      throwIfCancelled(ctx);
      if (op.kind === 'update') {
        const key = keyNameFrom(op.key, 'update');
        const value = valueFrom(op.changes, 'update');
        await assertEditableType(conn, key);
        await conn.set(key, value);
        affectedRows += 1;
      } else if (op.kind === 'delete') {
        // DEL is type-agnostic — works for any of the six redis types alike.
        const key = keyNameFrom(op.key, 'delete');
        const deleted = await conn.del(key);
        affectedRows += deleted;
      } else {
        // NX: creating a brand-new key must never silently overwrite an existing one — that's
        // what `update` (a plain SET) is for. A collision surfaces as a query-time condition,
        // not a connection failure (mirrors read.ts's "key no longer exists" precedent).
        const key = keyNameFrom(op.values, 'insert');
        const value = valueFrom(op.values, 'insert');
        const created = await conn.set(key, value, 'NX');
        if (created !== 'OK') throw new AdapterError('E_QUERY', `key already exists: ${key}`);
        affectedRows += 1;
      }
    }
  } catch (err) {
    if (err instanceof AdapterError) throw err;
    throw mapError(err);
  }

  return { affectedRows };
}

File: apps/kira-studio/internal/adapters/redis/mutate.go
package redis

import (
	"context"

	goredis "github.com/redis/go-redis/v9"

	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/adapters"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/storage/model"
)

// The reserved sentinels for redis mutations, expressed through the existing relational-shaped
// MutationRowOp rather than widening the shared mutation schema — mirrors mongo/mutate.go's
// $document precedent. keySentinel names the target redis key: plan.path only ever resolves to a
// database, never a specific key, so every op carries its own key name — including insert, which
// by definition has no existing key a path could point at yet.
const keySentinel = "_key"

// resolveDatabaseSegment is mutate's own single-database path check (P113 G2) — plan.path only ever
// resolves to a database, never a specific key (see keySentinel above), so RequirePath's exact-length
// match is safe here. Message text moves from "database-rooted path" to RequirePath's fixed
// "database path" wording — no test or frontend code pins the old string.
func resolveDatabaseSegment(path model.NodePath) (string, error) {
	segs, err := adapters.RequirePath(path, "mutate", adapters.Seg("database"))
	if err != nil {
		return "", err
	}
	return segs[0].Name, nil
}

func keyNameFrom(values model.RowValues, label string) (string, error) {
	raw, ok := values.Get(keySentinel)
	if !ok || raw == nil || *raw == "" {
		return "", adapters.New(adapters.CodeQuery, "a redis "+label+" mutation requires a non-empty "+keySentinel, nil)
	}
	return *raw, nil
}

func renderOpText(op model.MutationRowOp) (string, error) {
	switch op.Kind {
	case "update":
		key, err := keyNameFrom(op.Key, "update")
		if err != nil {
			return "", err
		}
		value, err := adapters.ValueFrom(op.Changes, "update")
		if err != nil {
			return "", err
		}
		// KEEPTTL: editing a value must not silently clear whatever expiry the key already had —
		// a plain SET (this adapter's own prior behaviour) resets TTL to none, which is a real
		// data-loss surprise on every edit of a key that happens to expire (P2 R1 finding).
		return "SET " + key + " " + value + " KEEPTTL", nil
	case "delete":
		key, err := keyNameFrom(op.Key, "delete")
		if err != nil {
			return "", err
		}
		return "DEL " + key, nil
	default: // insert
		key, err := keyNameFrom(op.Values, "insert")
		if err != nil {
			return "", err
		}
		value, err := adapters.ValueFrom(op.Values, "insert")
		if err != nil {
			return "", err
		}
		return "SET " + key + " " + value + " NX", nil
	}
}

// preview is mutate.ts's preview — synchronous (Adapter rule 3): no network, no TYPE lookup.
func preview(plan model.MutationPlan) ([]string, error) {
	if _, err := resolveDatabaseSegment(plan.Path); err != nil {
		return nil, err
	}
	out := make([]string, len(plan.Ops))
	for i, op := range plan.Ops {
		text, err := renderOpText(op)
		if err != nil {
			return nil, err
		}
		out[i] = text
	}
	return out, nil
}

// assertEditableType ports mutate.ts's assertEditableType: edit is scoped to string-type keys
// only (documented scope decision) — a hash/list/set/zset/stream each has its own per-element
// mutation semantics, a materially bigger job than a single SET.
func assertEditableType(ctx context.Context, conn *goredis.Client, key string) error {
	rawType, err := conn.Type(ctx, key).Result()
	if err != nil {
		return mapError(err)
	}
	if rawType != "none" && rawType != "string" {
		return adapters.New(adapters.CodeUnsupported, "only string-type keys are editable in this version, "+key+" is "+rawType, nil)
	}
	return nil
}

// applyUpdate is mutateDB's "update" arm.
func applyUpdate(ctx context.Context, conn *goredis.Client, rowOp model.MutationRowOp) (int, error) {
	key, err := keyNameFrom(rowOp.Key, "update")
	if err != nil {
		return 0, err
	}
	value, err := adapters.ValueFrom(rowOp.Changes, "update")
	if err != nil {
		return 0, err
	}
	if err := assertEditableType(ctx, conn, key); err != nil {
		return 0, err
	}
	// goredis.KeepTTL, not 0: a plain SET clears the key's existing expiry, silently
	// dropping a TTL the user never asked to change (P2 R1 finding).
	if err := conn.Set(ctx, key, value, goredis.KeepTTL).Err(); err != nil {
		return 0, mapError(err)
	}
	return 1, nil
}

// applyDelete is mutateDB's "delete" arm. DEL is type-agnostic — works for any of the six redis
// types alike.
func applyDelete(ctx context.Context, conn *goredis.Client, rowOp model.MutationRowOp) (int, error) {
	key, err := keyNameFrom(rowOp.Key, "delete")
	if err != nil {
		return 0, err
	}
	deleted, err := conn.Del(ctx, key).Result()
	if err != nil {
		return 0, mapError(err)
	}
	return int(deleted), nil
}

// applyInsert is mutateDB's default (insert) arm: NX — creating a brand-new key must never
// silently overwrite an existing one; that's what update (a plain SET) is for. A collision
// surfaces as a query-time condition, not a connection failure.
func applyInsert(ctx context.Context, conn *goredis.Client, rowOp model.MutationRowOp) (int, error) {
	key, err := keyNameFrom(rowOp.Values, "insert")
	if err != nil {
		return 0, err
	}
	value, err := adapters.ValueFrom(rowOp.Values, "insert")
	if err != nil {
		return 0, err
	}
	created, err := conn.SetNX(ctx, key, value, 0).Result()
	if err != nil {
		return 0, mapError(err)
	}
	if !created {
		return 0, adapters.New(adapters.CodeQuery, "key already exists: "+key, nil)
	}
	return 1, nil
}

// mutateDB is mutate.ts's mutate.
func mutateDB(ctx context.Context, conn *goredis.Client, op *adapters.OpCtx, readOnly bool, plan model.MutationPlan) (model.MutationResult, error) {
	// §8.12's standard: enforced here, not only greyed out in the UI.
	if err := adapters.AssertWritable(readOnly); err != nil {
		return model.MutationResult{}, err
	}
	if _, err := resolveDatabaseSegment(plan.Path); err != nil {
		return model.MutationResult{}, err
	}

	statements, err := preview(plan)
	if err != nil {
		return model.MutationResult{}, err
	}

	return adapters.RunKindDispatched(ctx, op, plan, readOnly, statements, adapters.DispatchUpdateDeleteInsert(
		func(ctx context.Context, rowOp model.MutationRowOp) (int, error) {
			return applyUpdate(ctx, conn, rowOp)
		},
		func(ctx context.Context, rowOp model.MutationRowOp) (int, error) {
			return applyDelete(ctx, conn, rowOp)
		},
		func(ctx context.Context, rowOp model.MutationRowOp) (int, error) {
			return applyInsert(ctx, conn, rowOp)
		},
	))
}
